Limit spec-defined option values (#929)
unicode-org/message-format-wg: spec/README.md

> (such as due to the release of new CLDR versions).

Updates to this specification will only reserve, define, or require
function identifiers and function option identifiers
which satisfy either of the following two requirements:
- Includes no namespace,

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9, and the characters '.', '-', and '_'.

Author: If we're going to reserve these, let's at least make it easy for a reader to be sure about which dots and dashes we mean.
```suggestion
  and has a name consisting of characters in the ranges a-z, A-Z, and 0-9,
  and the characters U+002E FULL STOP `.`, U+002D HYPHEN-MINUS `-`, and U+005F LOW LINE `_`.
```

> - Uses a namespace consisting of a single character in the range a-z.

Reviewer: Note well: names (and thus namespaces) are CASE SENSITIVE. We should reserve both a-z and A-Z as a result.

Author: ```suggestion
- Uses a namespace consisting of a single character in the ranges a-z and A-Z.
```

> (such as due to the release of new CLDR versions).

Updates to this specification will only reserve, define, or require
function identifiers and function option identifiers
which satisfy either of the following two requirements:
- Includes no namespace,

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Author: I don't think this is necessary?

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Reviewer 2: It gives us a bit more flexibility in the future, in case we want to use foo.bar in the future

I feel less strongly about '_'.

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Author: For function options we're already using camelCase, but I don't oppose reserving `.` as well.

@aphillips, if you agree, merge this suggestion?
```suggestion
  and has a name consisting of characters in the ranges a-z, A-Z, and 0-9,
  and the character U+002E FULL STOP `.`.
```

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Reviewer: I think I'd rather merge the first suggestion. Note that reserving `.`, `-`, and `_` does not prevent implementations or users from using these characters (this is what namespaces are for). A certain amount of flexibility is good for "future us". TBH, I think `.` is my least favorite, since it smacks of properties. The symbols don't meaning anything, of course.

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Reviewer 2: I would rather allow us to use '.' in the future, for future proofing. It doesn't prevent others from using that value.

>   and has a name consisting of characters in the ranges a-z, A-Z, and 0-9.

Reviewer: @macchiati Do you have an objection to the other two characters? I have no objection to `.`

- Uses the `u:` namespace.

All other identifiers in these categories are reserved for the use of implementations or users.

> Updates to this specification will only reserve, define, or require
> function option values
> consisting of characters in the ranges a-z, A-Z, and 0-9,
> and the characters U+002D HYPHEN-MINUS `-` and U+002E FULL STOP `.`.
> All other option values are reserved for the use of implementations or users.

Reviewer: I don't think this is quite right. The problem is that we already have some options that accept e.g. literal and we want to be somewhat permissive here (notice that you're carving out numeric literals). The real limitation wants to restrict our defining keywords.

Notice that the current policy has **no** promises on option _values_. I think it is hard to restrict "future us" from defining more keywords or special literal restrictions blindly. I don't think we should plunge into it this way. Instead, we should carve out what we know we won't do. Perhaps:

```suggestion
Updates to this specification will not reserve, define, or require
_option_ values in **standard** or **optional** _functions_ that start with the following list of symbols.
These values are reserved for use by implementations or user-defined extensions.
- `\_` LOW LINE
- _something else??_
```

Reviewer 2: I like reserving the u: namespace for Unicode. I would go a bit further in reserving all the single-letter namespaces at least for now, until we have a fuller picture of how to structure them to avoid collisions (both across organizations and across time).

However, I don't like the reservation of option value characters at all, based on the discussion.

> Updates to this specification will only reserve, define, or require
> function option values
> consisting of characters in the ranges a-z, A-Z, and 0-9,
> and the characters U+002D HYPHEN-MINUS `-` and U+002E FULL STOP `.`.
> All other option values are reserved for the use of implementations or users.

Reviewer: I think we agreed to remove this so

```suggestion
```
